File: java/slp/loop_deps.java

import java.util.Arrays;

public class loop_deps {
/*
  public static float micro1(float [] res1, float [] res2, float [] src1, float [] src2, int m, int n) {
      for (int i = 4; i < res1.length; i++) {
          // Inter statement dependency, forward write leading to failed vectorization.
          // Should be a no-dependency.
          res1[i] = src1[i] + src2[i];
          res2[i] = res1[i-4] * src2[i];
      }
      return res1[m] + res2[n];
  }
  public static float micro1(float [] res1, float [] res2, float [] src1, float [] src2, int m, int n) {
      for (int i = 4; i < res1.length; i++) {
          // Inter statement dependency, forward reads leading to failed vectorization.
          // This is correct.
          res1[i-4] = src1[i] + src2[i];
          res2[i] = res1[i] * src2[i];
      }
      return res1[m] + res2[n];
  }
*/
  public static float micro1(float [] res1, float [] res2, float [] src1, float [] src2, int m, int n) {
      for (int i = 4; i < res1.length; i++) {
          // Inter statement dependency, forward write leading to failed vectorization.
          // Should be a no-dependency.
          res1[i-4] = res1[i] + src2[i];
          res2[i] = res2[i] * src2[i];
      }
      return res1[m] + res2[n];
  }
  public static void main(String [] args) {
      float [] src1 = new float[1024]; 
      float [] src2 = new float[1024]; 
      float [] res1 = new float[1024]; 
      float [] res2 = new float[1024]; 
      Arrays.fill(src1, 10.0f);
      Arrays.fill(src2, 20.0f);
      for (int i = 0; i < 100000; i++) {
          micro1(res1, res2, src1, src2, 10, 20);
      }
      System.out.println("[res] " + Arrays.toString(Arrays.copyOfRange(res2, 0, 16)));
  }
}

File: java/slp/loop_deps.c

#if 0
float micro1(float * res1, float * res2, float * src1, float * src2, int length, int m, int n) {
    for (int i = 0; i < 1024; i++) {
        // loop carried inter statement dependency, limiting VF to 4.
        res1[i] = src1[i] + src2[i];
        res2[i] = res1[i+4] + src2[i];
    }
    return res1[m] + res2[n];
}
#endif

#if 0
float micro2(float * res1, float * res2, float * src1, float * src2, int length, int m, int n) {
    for (int i = 0; i < 1024; i++) {
        // Forward reads intra-statment not a problem, Anti-dependency
        res1[i] = res1[i + 4] + src2[i];
    }
    return res1[m] + res2[n];
}
#endif

#if 0
float micro3(float * res1, float * res2, float * src1, float * src2, int length, int m, int n) {
    for (int i = 4; i < 1024; i++) {
        // forward write intra-statment, true dependence raw, limiting vf to 4.
        res1[i] = res1[i - 4] + src2[i];
    }
    return res1[m] + res2[n];
}
#endif

#if 0
float micro4(float * res1, float * res2, float * src1, float * src2, int length, int m, int n) {
    for (int i = 0; i < 1024; i++) {
        // forward write intra-statment, true dependence raw, limiting vf to 4.
        res1[i + 4] = res1[i] + src2[i];
    }
    return res1[m] + res2[n];
}
#endif

float micro5(float * res1, float * res2, float * src1, float * src2, int length, int m, int n) {
    for (int i = 4; i < 1024; i++) {
        // interstatement, forward read, limits VF to 4..
        res1[i-4] = src1[i] + src2[i];
        res2[i] = res1[i] * src2[i];
    }
    return res1[m] + res2[n];
}
